### PR TITLE
Add List monad

### PR DESCRIPTION
**Pull request by flash-gordon**
This adds the `List` monad. This will be a nice addition to `Either` and `Maybe` for things like lists of operations.

This is a WIP, I'm going to add more stuff related to integration with other types of monads.

/cc @solnic @timriley

### CODE REVIEW

File: lib/dry/monads/list.rb [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+      #
+      # @return [String]
+      def inspect
+        "List#{ value.inspect }"
+      end
+      alias_method :to_s, :inspect
+
+      # Coerces to an array
+      alias_method :to_ary, :value
+      alias_method :to_a, :to_ary
```
Comment (solnic) [inline lib/dry/monads/list.rb]: Is it supposed to quack like an array? `to_ary` should exist only if an object is array-like (and yes, I made that mistake in rom, where `Relation#to_ary` is implemented, but relations *are not array-like* so it has to be fixed)

Comment (flash-gordon) [inline lib/dry/monads/list.rb]: well, I'm not _sure_ for 100%, `List` is an array inside, the value will always be an `Array` instance, so they are defo very close to each other. But `List` won't offer any mutation, is it enough to not have `#to_ary`? Dunno. `#first` and `#last` also return `Maybe`s, but I think they should be replaced with `#head` and `#tail`

Comment (solnic) [inline lib/dry/monads/list.rb]: If an object has `to_ary` then it means you could use it *as an array*, which means if there's code relying on array objects, it should work with list instances too, so I'd say we shouldn't have `#to_ary`. FWIW `Kernel.Array` coercion function works with objects that have `to_a` so we should be OK

Comment (flash-gordon) [inline lib/dry/monads/list.rb]: having `#to_ary` allows to have some niceties like destructuring

Comment (flash-gordon) [inline lib/dry/monads/list.rb]: @solnic it's not completely true, `#to_ary` means that an object acts as an array _in many aspects_, but not all of them
Consider `Float#to_int`, it means floats act as integers, but they are still _different, that is you can't call `Float#gcd`, `gcd` is defined only for integers. If your code expects an integer, you should call `#to_int`

Comment (flash-gordon) [inline lib/dry/monads/list.rb]: Unfortunately I didn't find an example of implementing `#to_ary` by other class than `Array`  in ruby itself

**Comment by flash-gordon** [comment]
I added `Monad#fmap2` and `Monad#fmap3` so that we can do this
```ruby
List[Right(1), Left(1)].fmap2 { |x| x + 1 } # => List[Right(2), Left(1)]
```
and this 
```ruby
Right(Some(1)).fmap2 { |x| x + 1 } # => Right(Some(2))
```

I'll merge the branch now, but I still need to figure out what to do with `traverse` before releasing the gem. I'll send another PR for that.

**Pull request merged**